ipa-core/src/protocol/ipa_prf/boolean_ops/share_conversion_aby.rs

```diff
@@ -126,84 +132,103 @@ where
     );
 
     // Ensure that the probability of leaking information is less than 1/(2^128).
-    debug_assert!(x.iter().count() < (BITS - 128));
-
-    // generate sh_r = (0, 0, sh_r) and sh_s = (sh_s, 0, 0)
-    // the two highest bits are set to 0 to allow carries for two additions
-    let (sh_r, sh_s) =
-        gen_sh_r_and_sh_s::<_, BITS, NC>(&ctx.narrow(&Step::GenerateSecretSharing), record_id);
-
-    // generate upgraded context
-    // we expect 2*256 = 512 gates in total for two additions
-    // the vectorization factor is NC
-    // the total amount of multiplications is therefore NC*512
-    // make sure proof does not get too big, i.e. less that 100k
-    // (such that there are less than 50m multiplications)
-    assert!(NC < 100_000);
-    let validator = ctx.dzkp_validator(NC);
-    let m_ctx = validator.context();
-
-    // addition r+s might cause carry,
-    // this is no problem since we have set bit 254 of sh_r and sh_s to 0
-    let sh_rs = {
-        let (mut rs_with_higherorderbits, _) = integer_add::<_, TwoHundredFiftySixBitOpStep, NC>(
-            m_ctx.narrow(&Step::IntegerAddBetweenMasks),
-            record_id,
-            &sh_r,
-            &sh_s,
+    debug_assert!(input_shares.first().iter().count() < (BITS - 128));
```

Comment:
I don't think this is doing what it may seem to do: https://doc.rust-lang.org/std/option/struct.Iter.html

```rust
    #[test]
    fn test_option_iter() {
        let x = BitDecomposed::new(repeat_n(AdditiveShare::<BA64>::ZERO, 256));
        let bd = vec![x.clone(), x];
        debug_assert!(bd.first().iter().count() < (256 - 128)); // pass
        debug_assert!(bd[0].iter().count() < (256 - 128)); // fail
    }

```

We should add a test that checks it

Reply:
This is a good catch, thank you.